framework/audio/engine/internal/audioengine.h

```diff
@@ -47,29 +49,50 @@ class AudioEngine : public IAudioEngine
 
     handle play(std::shared_ptr<IAudioSource> s, float volume = -1, float pan = 0, bool paused = false) override;
     void seek(handle h, time sec) override;
-    void stop(handle h) override;
     void setPause(handle h, bool paused) override;
+    void stop(handle h) override;
 
-    void stopAll();
+    Status status(handle h) const override;
+    async::Channel<Status> statusChanged(handle h) const override;
 
     time position(handle h) const override;
-    bool isEnded(handle h) const override;
 
     void setVolume(handle h, float volume) override;
     void setPan(handle h, float val) override;
     void setPlaySpeed(handle h, float speed) override;
 
-    async::Notification playCallbackCalled() const override;
-
-    // internal
-    void onPlayCallbackCalled();
+    void swapPlayContext(handle h, Context& ctx) override;
```

Comment:
Can you describe the case when we'll need to swap the context?

Reply:
When the audio source is played, the context is placed in the audioengine that matches the current audio data.
This is done frequently, for each call to the callback driver (about 30 ms).
The source itself no longer needs this context.
Therefore, we can perform optimization without copying the context (and its data) from the source to audioengine, but swap contexts is much more efficient.